dexc-desktop: Snap package

peterzen opened the pull request:
This PR adds configuration, build scripts and Github workflow to build the Snap package and publishing it in the [Snap Store](https://snapcraft.io/store).

The snap build uses the .deb package.  The PR also contains a few fixes/improvements in the deb build:

 *   Icons: added SVG icon for hicolor and symbolic variations, fixed 128px PNG to match official version
 *   icon install location changed to /usr/share/icons which is the XDG standard
 *  binary installs in /usr/bin directly (this is standard practice, simplifies deb scripts)
 *   added AppStream metainfo file (for GNOME/KDE/Ubuntu software centers and the snap)

JoeGruffins reviewed (commented):
Getting an error when running `./pkg/pkg-debian.sh`:
```
dpkg-deb: error: parsing file './build/dexc_bf46fe348fb83893408e875408499761a3df5ea4-0_amd64/DEBIAN/control' near line 2 package 'dexc':
 'Version' field value 'bf46fe348fb83893408e875408499761a3df5ea4': version number does not start with digit
```

JoeGruffins commented on client/cmd/dexc-desktop/README.md:
````diff
@@ -7,11 +7,29 @@ The dexc-desktop client is a cross-platform desktop application for the DEX clie
 
-For development, you'll need to install the WebKit Development Libraries.
-`apt install libgtk-3-dev libwebkit2gtk-4.0-dev`
+For development, you'll need to install the WebKit Development Libraries and the
+standard build toolchain:
+
+```bash
+apt install libgtk-3-dev libwebkit2gtk-4.0-dev build-essential
+```
````
I tried this on a fresh installation and I think you need to build the site first at `client/webserver/site` as well? Not sure if worth mentioning here.

peterzen replied:
Yes, good point - in development `site/` needs to be built indeed, on the release branch it's not necessary.

peterzen commented:
> dpkg-deb: error: parsing file './build/dexc_bf46fe348fb83893408e875408499761a3df5ea4-0_amd64/DEBIAN/control' near line 2 package 'dexc':
>  'Version' field value 'bf46fe348fb83893408e875408499761a3df5ea4': version number does not start with digit
> ```

It pulls the version number from the `git` tag - the idea being that for release builds there's always a tag but for dev builds  the commit hash confuses the deb build.  Maybe this needs to be reconsidered.

JoeGruffins commented:
> It pulls the version number from the git tag - the idea being that for release builds there's always a tag but for dev builds the commit hash confuses the deb build. Maybe this needs to be reconsidered.

So the commit needs a tag to work? I did not know. Maybe add that to the readme there at least if it is a requirement?

JoeGruffins reviewed (approved):
After adding a bogus tag was able to build and install with snap.

buck54321 commented on client/cmd/dexc-desktop/metainfo/org.decred.dcrdex.metainfo.xml:
```diff
@@ -0,0 +1,30 @@
+<component type="desktop-application">
+  <id>org.decred.dcrdex</id>
+  <launchable type="desktop-id">org.decred.dcrdex.desktop</launchable>
+  <name>DCRDEX</name>
```
Decred DEX Client

buck54321 commented on client/cmd/dexc-desktop/metainfo/org.decred.dcrdex.metainfo.xml:
```diff
@@ -0,0 +1,30 @@
+  <id>org.decred.dcrdex</id>
+  <launchable type="desktop-id">org.decred.dcrdex.desktop</launchable>
+  <name>DCRDEX</name>
+  <summary>Decentralized cryptocurrency exchange</summary>
```
Multi-wallet backed by Decred DEX

buck54321 commented on client/cmd/dexc-desktop/metainfo/org.decred.dcrdex.metainfo.xml:
```diff
@@ -0,0 +1,30 @@
+  <launchable type="desktop-id">org.decred.dcrdex.desktop</launchable>
+  <name>DCRDEX</name>
+  <summary>Decentralized cryptocurrency exchange</summary>
+  <metadata_license>MIT</metadata_license>
```
You'd think if you specified an MIT license here, we would need to have an MIT license somwhere. Can this just be `BlueOak-1.0.0` too, for consistency?

peterzen replied:
https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-metadata_license

buck54321 commented on client/cmd/dexc-desktop/metainfo/org.decred.dcrdex.metainfo.xml:
```diff
@@ -0,0 +1,30 @@
+  <developer_name>The Decred Developers</developer_name>
+  <description>
+    <p>
+      Non-custodial peer-to-peer exchange where your funds never leave your wallet.  
+      Trade Bitcoin, USDC, Ethereum, Decred, Dogecoin, Zcash and more without KYC 
+      and fees, using atomic swaps technology.
```
~~Non-custodial peer-to-peer exchange where your funds never leave your wallet.~~
Self-custodial multi-wallet with built-in peer-to-peer exchange.

buck54321 commented on client/cmd/dexc-desktop/metainfo/org.decred.dcrdex.metainfo.xml:
```diff
@@ -0,0 +1,30 @@
+  </description>
+  <screenshots>
+    <screenshot type="default">
+      <image>https://raw.githubusercontent.com/decred/umbrel-app-store/master/decred-dcrdex/images/screenshot_1.png</image>
+    </screenshot>
+    <screenshot>
+      <image>https://raw.githubusercontent.com/decred/umbrel-app-store/master/decred-dcrdex/images/screenshot_2.png</image>
+    </screenshot>
+    <screenshot>
+      <image>https://raw.githubusercontent.com/decred/umbrel-app-store/master/decred-dcrdex/images/screenshot_3.png</image>
```
These images are really outdated. Can we host the images in our own repo?

peterzen replied:
Yes, these are meant to be placeholder images only.  We need to take the screenshots whenever appropriate and the URLs can point to the images in repo.

buck54321 commented on client/cmd/dexc-desktop/pkg/common.sh:
```diff
@@ -0,0 +1,19 @@
+# in this directory.
+
+# pick up the release tag from git
+VER=$(git describe --tags --abbrev=0 --always | sed -e 's/^v//')
```
Wow! I need to do some testing, but this is slick.

peterzen replied:
The caveat with this one, though, is that it requires a tag on the repo clone being built (JoeGruff tripped over this above).  So for release builds it'll work but it won't work when only a commit hash is available due to the debian package builder requiring a version format x.y.z

buck54321 reviewed (changes requested):
I'm not seeing the benefit of the github actions deployment. What does it do for us over just running the script ourselves?

> Icons: added SVG icon for hicolor and symbolic variations, fixed 128px PNG to match official version

Can you say more about the new files and what they do for us? Or point me to some documentation? 

I don't care about the official logo version. I changed it because the official color was too dark and looked bad in my panel. None of this matters much, of course, because we're rebranding anyway, but prefer to use the lighter color that has margins.

peterzen commented:
> I'm not seeing the benefit of the github actions deployment. What does it do for us over just running the script ourselves?
> 
It does the same thing but makes life a tad easier.  `snapcraft`, the snap builder isn't straightforward to run on systems other than Ubuntu as it requires several manual steps to initialize `lxd` which the build relies on.  So basically you need to run the script in an Ubuntu VM which I presumed isn't the system of choice for running release builds.

The binary it packages into a snap comes from the .deb package.

> > Icons: added SVG icon for hicolor and symbolic variations, fixed 128px PNG to match official version
> 
> Can you say more about the new files and what they do for us? Or point me to some documentation?
> 

dexc.png is scaled to 128px which is one of the default icon sizes, the 100px version was scaled up and rendered blurry
dexc.svg is a scalable version which the desktop environment picks up when a bitmap in the required size is not available
dexc-symbolic.svg is a B/W version that the desktop environment might use with certain accessibility settings like monochrome etc.  Also this is ideally the icon to be used in the systray as in that size this provides the most contrast.

https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-latest.html

> I don't care about the official logo version. I changed it because the official color was too dark and looked bad in my panel. None of this matters much, of course, because we're rebranding anyway, but prefer to use the lighter color that has margins.

The tray icon should really match the other icons (menu, dock, task switcher), if it's different it looks odd.  How about using the monochrome version?  That's the most distinctly recognizable in the tray visually.

buck54321 commented on client/cmd/dexc-desktop/pkg/pkg-debian.sh:
```diff
@@ -54,18 +39,19 @@ DOT_DESKTOP_BUILDDIR="${DEB_DIR}${DOT_DESKTOP_TARGETDIR}"
-mkdir -p "${ICON_BUILDDIR}"
+
+# check if site bundle is built, if not build it
+if [ ! -d ../../webserver/site/dist ]; then
+  CWD=$(pwd)
+  cd ../../webserver/site
+  npm clean-install
+  npm run build
+  cd $CWD
+fi
```
Just because there's a file there doesn't mean its the right one. Any reason not to just build it every time?

peterzen replied:
The idea here was that the site package is added to the repo on the release branch so there's no need to build it at the time of building the release packages but it certainly won't hurt.

buck54321 commented on client/cmd/dexc-desktop/pkg/common.sh:
```diff
@@ -0,0 +1,19 @@
+# in this directory.
+
+# pick up the release tag from git
+VER=$(git describe --tags --abbrev=0 --always | sed -e 's/^v//')
```
Hmm. If I do

```Bash
git fetch upstream --tags
git checkout -b v0.6.3 v0.6.3
git describe --tags --abbrev=0 --always | sed -e 's/^v//'
```
I get
```Bash
0.6.2
```

peterzen replied:
Does `upstream` point to `github.com/decred/dcrdex`?  I'm getting the correct tag on a fresh clone.

```bash
$ git clone https://github.com/decred/dcrdex
Cloning into 'dcrdex'...
...
$ cd dcrdex
$ git fetch origin --tags
$ git checkout -b v0.6.3 v0.6.3
Switched to a new branch 'v0.6.3'
$ git describe --tags --abbrev=0 --always | sed -e 's/^v//'
0.6.3
```

buck54321 commented on client/cmd/dexc-desktop/pkg/pkg-debian.sh:
```diff
@@ -83,40 +69,39 @@ Depends: libgtk-3-0, libwebkit2gtk-4.0-37
 cat > "${DOT_DESKTOP_BUILDPATH}" <<EOF
 [Desktop Entry]
-Version=${VER}
+Version=1.0
```
Maybe 1.5?

buck54321 commented on client/cmd/dexc-desktop/pkg/publish-snap.sh:
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+
+source $(dirname "$0")/common.sh
+
+snapcraft login
+
```
This will prompt for snapcraft account credentials? Is it expected that the app is already `snapcraft register`'d?

peterzen replied:
Yes, it will ask for the account creds - the app must be registered in the same account.
https://snapcraft.io/snaps

buck54321 commented on client/cmd/dexc-desktop/snap/local/snapcraft.yaml.in:
```diff
@@ -0,0 +1,102 @@
+      # Point icon to the correct location
+      sed -i -e 's|Icon=.*$|Icon=usr/share/icons/hicolor/scalable/apps/dexc.svg|' $SNAPCRAFT_PART_INSTALL/usr/share/applications/dexc.desktop
+  
+    prime:
+      - -usr/lib/x86_64-linux-gnu/libEGL_mesa*
+      - -usr/lib/x86_64-linux-gnu/libGLX_mesa*
+      - -usr/lib/x86_64-linux-gnu/libGLESv2*
+      - -usr/lib/x86_64-linux-gnu/libcaca++*
+      - -usr/lib/x86_64-linux-gnu/libcolordprivate*
+      - -usr/lib/x86_64-linux-gnu/libdconf*
+      - -usr/lib/x86_64-linux-gnu/libexslt*
+      - -usr/lib/x86_64-linux-gnu/libgstcheck-1.0*
+      - -usr/lib/x86_64-linux-gnu/libgstcontroller-1.0*
+      - -usr/lib/x86_64-linux-gnu/libicuio*
+      - -usr/lib/x86_64-linux-gnu/libicutest*
+      - -usr/lib/x86_64-linux-gnu/libjacknet*
+      - -usr/lib/x86_64-linux-gnu/libjackserver*
+      - -usr/lib/x86_64-linux-gnu/liborc-test-0.4*
+      - -usr/lib/x86_64-linux-gnu/libpulse-simple*
+      - -usr/lib/x86_64-linux-gnu/libunwind-coredump*
+      - -usr/lib/x86_64-linux-gnu/libunwind-ptrace*
+      - -usr/lib/x86_64-linux-gnu/libunwind-x86_64*
+      - -usr/lib/x86_64-linux-gnu/libwoff2enc*
+      - -usr/lib/x86_64-linux-gnu/libicutu*
```
How is this list generated and what does it mean?

peterzen replied:
These are libraries that .deb packages in the `dexc.deb` dependency tree (`libgtk-3-0`, `libwebkit2gtk-4.0-37` and their deps) install but are not required.  The list is emitted by `snapcraft`.

buck54321 commented on client/cmd/dexc-desktop/snap/local/snapcraft.yaml.in:
```diff
@@ -0,0 +1,102 @@
+      #TMPDIR: $XDG_RUNTIME_DIR
+      #XDG_CURRENT_DESKTOP: Unity
+      #DISABLE_WAYLAND: 1
+    plugs: 
+      - home
+      - opengl
+      - x11
+      - desktop
+      - desktop-legacy
+      - network
+      - network-status
+      - browser-support
+      - screen-inhibit-control
+      - dbus-svc
+      - dbus-dbusmenu
+      - dbus-statusnotifier
```
How is this list generated and what does it mean?

peterzen replied:
This list defines the snap interfaces the app needs access to, basically they translate into AppArmor permissions.  It's assembled from the documentation and trial-and-error.

https://snapcraft.io/docs/interface-management

buck54321 commented on client/cmd/dexc-desktop/go.mod:
```diff
@@ -66,7 +66,7 @@ require (
 	github.com/decred/dcrd/dcrec/secp256k1/v4 v4.2.0 // indirect
 	github.com/decred/dcrd/dcrjson/v4 v4.0.1 // indirect
-	github.com/decred/dcrd/dcrutil/v4 v4.0.1 // indirect
+	github.com/decred/dcrd/dcrutil/v4 v4.0.2-0.20231005000813-c102e54b4128 // indirect
```
Did we need this for something?

peterzen replied:
This is to incorporate https://github.com/decred/dcrd/pull/3196

buck54321 commented on .github/workflows/README.md:
```diff
@@ -0,0 +1,7 @@
+
+This workflow builds a `snap` package and uploads it to the [Snap Store](https://snapcraft.io/store).  It has a `workflow_dispatch` trigger so it can be triggered manually only, it is not hooked up to any other GH event.
+
+The store upload requires the `SNAPCRAFT_STORE_CREDENTIALS` variable to be set in [Github Secrets](https://github.com/decred/dcrdex/settings/secrets/actions).  
+See https://github.com/snapcore/action-publish for details.
```
The https://github.com/decred/dcrdex/settings/secrets/actions link is busted and I don't have the "Settings" tab as described in action-publish readme.

peterzen replied:
You need to be the owner of the repo to be able to access the settings.

buck54321 commented on .github/workflows/build-snap.yml:
```diff
@@ -0,0 +1,48 @@
+            SNAPCRAFT_STORE_CREDENTIALS: ${{ secrets.SNAPCRAFT_STORE_CREDENTIALS }}
+          with:
+            snap: ${{ steps.build.outputs.snap }}
+            release: stable
```
I believe if this is omitted, the package is uploaded but not released. Would that give us a chance to double-check everything first and then release later?

peterzen replied:
It would, though one could also run the snap build locally to test the package.

JoeGruffins commented:
The icon [client/cmd/bisonw-desktop/src/bisonw.svg](https://github.com/decred/dcrdex/pull/2580/files#diff-9b26e21a30254adb46cb5ef4c1858c68186ed57215ae1ecbf32b3ffc87ec5a48) looks outdated. Should be one of these top two maybe? https://github.com/decred/dcrdex/tree/master/docs/branding

JoeGruffins commented:
Trying to build on a clean ubuntu 24.04 but stuck at these dependencies.

````
E: Unable to locate package libwebkit2gtk-4.0-dev
E: Couldn't find any package by glob 'libwebkit2gtk-4.0-dev'
E: Couldn't find any package by regex 'libwebkit2gtk-4.0-dev'
```

peterzen commented on client/cmd/bisonw-desktop/README.md:
````diff
@@ -5,14 +5,44 @@ bison-desktop is a cross-platform desktop application for Bison Wallet. The inst
+
+### Preparing the build environment
+```bash
+# on Ubuntu:
+sudo apt-get install snap  
+# on Debian: 
+sudo apt-get install snapd
+sudo snap install --classic snapcraft
+sudo adduser $USER lxd
+newgrp lxd
+lxd init --auto
+sudo ufw disable # if ufw is installed on the host, this is neccessary for lxd to have network access
````
Yeah some things have changed since this was written 2 years ago... I wonder if it's worth warming up this PR at all, considering it wasn't merged for so long.

peterzen commented:
> E: Unable to locate package libwebkit2gtk-4.0-dev
> E: Couldn't find any package by glob 'libwebkit2gtk-4.0-dev'
> E: Couldn't find any package by regex 'libwebkit2gtk-4.0-dev'
> ```

Apparently, `libwebkit2gtk-4.0` was dropped from Ubuntu as of 24.0, but `github.com/webview/webview` has not been updated to link against 4.1 - this breaks the build (not just the snap package).  There's an open PR fixing this (https://github.com/webview/webview_go/pull/62/files) but it hasn't been merged for quite some time (webview_go seems to be unmaintained).  

We have 2 options - wait for that above PR to be merged, or fork the webview_go and use that going forward.  I think option 2 is the way to go.

JoeGruffins commented:
I guess this should be good to go after rebase?

JoeGruffins commented:
@peterzen sorry, I merged your other pr first and it created a conflict

JoeGruffins commented:
I was able to build on newest Ubuntu.
